### ISSUE TITLE
Support upcoming F# 4.1

### ISSUE THREAD

**Issue by majocha** [JamesNK/Newtonsoft.Json#1180]
There is a [breaking change](https://github.com/Microsoft/visualfsharp/issues/2257) in F# 4.1 concerning Json.NET.
It causes an exception when deserializing to FSharpList type

_"Newtonsoft.Json.JsonSerializationException: Cannot create and populate list type Microsoft.FSharp.Collections.FSharpList`1[System.Int32]. Path '', line 1, position 1."_

Because in F# 4.1 FSharp list type now [implements IReadOnlyCollection](https://github.com/Microsoft/visualfsharp/commit/13b0ba050d450ca09a3e461d6355e490a273fca9)

this conditional block in [JsonArrayContract.cs](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L182)
```csharp
#if HAVE_READ_ONLY_COLLECTIONS
else if (ReflectionUtils.ImplementsGenericDefinition(underlyingType, typeof(IReadOnlyCollection<>), out tempCollectionType))
```
is executed instead of [the following](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs#L210)
```csharp
#if HAVE_FSHARP_TYPES
if (!HasParameterizedCreatorInternal && underlyingType.Name == FSharpUtils.FSharpListTypeName)
```

Basically as for now the logic in JsonArrayContract assumes FSharpList type doesn't implement IReadOnlyCollection. This is no longer true in the upcoming F# release.

**Comment by draganjovanovic1** [comment]
The issue is no more present in Newtonsoft.Json 10.*, FSharpList deserialization works well. Any chance to release 9.0.2 with the fix? I have a several libraries in the project having dependency binding on Newtonsoft.Json >= 9 & < 10. Thanks.

**Comment by JamesNK** [comment]
> Any chance to release 9.0.2 with the fix?

Zero

**Comment by draganjovanovic1** [comment]
Ok. Thanks anyway. Any known issues or incompatibility if I upgrade to 10.0.2?
